# Лабораторная работа № 2. Структуры данных

## Раздел 2.2. Повтор заданий.

### Кортежи

In [1]:
# пустой кортеж:
()

()

In [2]:
# кортеж из элементов типа String:
favoritelang = ("Python","Julia","R")

("Python", "Julia", "R")

In [3]:
# кортеж из целых чисел:
x1 = (1, 2, 3)

(1, 2, 3)

In [7]:
# кортеж из элементов разных типов:
x2 = (1, 2.0, "tmp")
#typeof(x2);

(1, 2.0, "tmp")

In [8]:
# именованный кортеж:
x3 = (a=2, b=1+2)

(a = 2, b = 3)

Примеры операций над кортежами:

In [9]:
# длина кортежа x2:
length(x2)

3

In [10]:
# обратиться к элементам кортежа x2:
x2[1], x2[2], x2[3]

(1, 2.0, "tmp")

In [11]:
# произвести какую-либо операцию (сложение)
# с вторым и третьим элементами кортежа x1:
с = x1[2] + x1[3]

5

In [12]:
# обращение к элементам именованного кортежа x3:
x3.a, x3.b, x3[2]

(2, 3, 3)

In [14]:
# проверка вхождения элементов tmp и 0 в кортеж x2
# (два способа обращения к методу in()):
in("tmp", x2), 0 ∈ x2

(true, false)

### Словари

Примеры словарей и операций над ними:

In [15]:
# создать словарь с именем phonebook:
phonebook = Dict("Иванов И.И." => ("867-5309","333-5544"), "Бухгалтерия" => "555-2368")

Dict{String, Any} with 2 entries:
  "Бухгалтерия" => "555-2368"
  "Иванов И.И." => ("867-5309", "333-5544")

In [16]:
# вывести ключи словаря:
keys(phonebook)

KeySet for a Dict{String, Any} with 2 entries. Keys:
  "Бухгалтерия"
  "Иванов И.И."

In [17]:
# вывести значения элементов словаря:
values(phonebook)

ValueIterator for a Dict{String, Any} with 2 entries. Values:
  "555-2368"
  ("867-5309", "333-5544")

In [18]:
# вывести заданные в словаре пары "ключ - значение":
pairs(phonebook)

Dict{String, Any} with 2 entries:
  "Бухгалтерия" => "555-2368"
  "Иванов И.И." => ("867-5309", "333-5544")

In [21]:
# проверка вхождения ключа в словарь:
haskey(phonebook, "Иванов И.И."), haskey(phonebook, "Иванов И.Д.")

(true, false)

In [23]:
# добавить элемент в словарь:
phonebook["Сидоров П.С."] = "555-3344"

"555-3344"

In [25]:
# удалить ключ и связанные с ним значения из словаря
pop!(phonebook, "Иванов И.И.")

("867-5309", "333-5544")

In [26]:
# Объединение словарей (функция merge()):
a = Dict("foo" => 0.0, "bar" => 42.0);
b = Dict("baz" => 17, "bar" => 13.0);
merge(a, b), merge(b,a)

(Dict{String, Real}("bar" => 13.0, "baz" => 17, "foo" => 0.0), Dict{String, Real}("bar" => 42.0, "baz" => 17, "foo" => 0.0))

### Множества

In [27]:
# создать множество из четырёх целочисленных значений:
A = Set([1, 3, 4, 5])

Set{Int64} with 4 elements:
  5
  4
  3
  1

In [28]:
# создать множество из 11 символьных значений:
B = Set("abrakadabra")

Set{Char} with 5 elements:
  'a'
  'd'
  'r'
  'k'
  'b'

In [29]:
# проверка эквивалентности двух множеств:
S1 = Set([1,2]);
S2 = Set([3,4]);
issetequal(S1,S2)

false

In [30]:
S3 = Set([1,2,2,3,1,2,3,2,1]);
S4 = Set([2,3,1]);
issetequal(S3,S4)

true

In [31]:
# объединение множеств:
C=union(S1,S2)

Set{Int64} with 4 elements:
  4
  2
  3
  1

In [32]:
# пересечение множеств:
D = intersect(S1,S3)

Set{Int64} with 2 elements:
  2
  1

In [33]:
# разность множеств:
E = setdiff(S3,S1)

Set{Int64} with 1 element:
  3

In [34]:
# проверка вхождения элементов одного множества в другое:
issubset(S1,S4)

true

In [35]:
# добавление элемента в множество:
push!(S4, 99)

Set{Int64} with 4 elements:
  2
  99
  3
  1

In [36]:
# удаление последнего элемента множества:
pop!(S4)

2

### Массивы

In [37]:
# создание пустого массива с абстрактным типом:
empty_array_1 = []

Any[]

In [40]:
# создание пустого массива с конкретным типом:
empty_array_2 = (Int64)[]
empty_array_3 = (Float64)[]

Float64[]

In [41]:
# вектор-столбец:
a = [1, 2, 3]

3-element Vector{Int64}:
 1
 2
 3

In [42]:
# вектор-строка:
b = [1 2 3]

1×3 Matrix{Int64}:
 1  2  3

In [45]:
# многомерные массивы (матрицы):
A = [[1, 2, 3] [4, 5, 6] [7, 8, 9]]

3×3 Matrix{Int64}:
 1  4  7
 2  5  8
 3  6  9

In [46]:
B = [[1 2 3]; [4 5 6]; [7 8 9]]

3×3 Matrix{Int64}:
 1  2  3
 4  5  6
 7  8  9

In [50]:
# одномерный массив из 8 элементов (массив $1 \times 8$)
# со значениями, случайно распределёнными на интервале [0, 1):
c = rand(1, 8)

1×8 Matrix{Float64}:
 0.679921  0.340091  0.821479  0.729865  …  0.343236  0.891298  0.400566

In [52]:
# многомерный массив $2 \times 3$ (2 строки, 3 столбца) элементов
# со значениями, случайно распределёнными на интервале [0, 1):
C = rand(2,3)

2×3 Matrix{Float64}:
 0.26461    0.555366  0.0877469
 0.0027883  0.635858  0.1202

In [53]:
# трёхмерный массив:
D = rand(4, 3, 2)

4×3×2 Array{Float64, 3}:
[:, :, 1] =
 0.209757  0.148723  0.945074
 0.371645  0.298075  0.576505
 0.272831  0.610135  0.891197
 0.911689  0.798986  0.378871

[:, :, 2] =
 0.874182  0.992663  0.486733
 0.074272  0.106636  0.441866
 0.615368  0.453661  0.41649
 0.431464  0.24774   0.572056

Примеры массивов, заданных некоторыми функциями через включение:

In [54]:
# массив из квадратных корней всех целых чисел от 1 до 10:
roots = [sqrt(i) for i in 1:10]

10-element Vector{Float64}:
 1.0
 1.4142135623730951
 1.7320508075688772
 2.0
 2.23606797749979
 2.449489742783178
 2.6457513110645907
 2.8284271247461903
 3.0
 3.1622776601683795

In [55]:
# массив с элементами вида 3*x^2,
# где x - нечётное число от 1 до 9 (включительно)
ar_1 = [3*i^2 for i in 1:2:9]

5-element Vector{Int64}:
   3
  27
  75
 147
 243

In [56]:
# массив квадратов элементов, если квадрат не делится на 5 или 4:
ar_2=[i^2 for i=1:10 if (i^2%5!=0 && i^2%4!=0)]

4-element Vector{Int64}:
  1
  9
 49
 81

In [57]:
# одномерный массив из пяти единиц:
ones(5)

5-element Vector{Float64}:
 1.0
 1.0
 1.0
 1.0
 1.0

In [58]:
# двумерный массив 2x3 из единиц:
ones(2,3)

2×3 Matrix{Float64}:
 1.0  1.0  1.0
 1.0  1.0  1.0

In [59]:
# одномерный массив из 4 нулей:
zeros(4)

4-element Vector{Float64}:
 0.0
 0.0
 0.0
 0.0

In [60]:
# заполнить массив 3x2 цифрами 3.5
fill(3.5,(3,2))

3×2 Matrix{Float64}:
 3.5  3.5
 3.5  3.5
 3.5  3.5

In [63]:
# заполнение массива посредством функции repeat():
repeat([1,2],3,3)

6×3 Matrix{Int64}:
 1  1  1
 2  2  2
 1  1  1
 2  2  2
 1  1  1
 2  2  2

In [289]:
repeat([1 2],3,3)

3×6 Matrix{Int64}:
 1  2  1  2  1  2
 1  2  1  2  1  2
 1  2  1  2  1  2

In [64]:
# преобразование одномерного массива из целых чисел от 1 до 12
# в двумерный массив 2x6
a = collect(1:12)
b = reshape(a,(2,6))

2×6 Matrix{Int64}:
 1  3  5  7   9  11
 2  4  6  8  10  12

In [65]:
# транспонирование
b'

6×2 adjoint(::Matrix{Int64}) with eltype Int64:
  1   2
  3   4
  5   6
  7   8
  9  10
 11  12

In [66]:
# транспонирование
с = transpose(b)

6×2 transpose(::Matrix{Int64}) with eltype Int64:
  1   2
  3   4
  5   6
  7   8
  9  10
 11  12

In [67]:
# массив 10x5 целых чисел в диапазоне [10, 20]:
ar = rand(10:20, 10, 5)

10×5 Matrix{Int64}:
 18  19  11  16  15
 12  16  11  17  12
 12  19  20  12  14
 11  16  20  15  12
 14  18  10  15  18
 17  18  16  17  12
 19  13  18  10  13
 18  11  12  12  18
 12  20  17  20  11
 16  19  10  20  19

In [68]:
# выбор всех значений строки в столбце 2:
ar[:, 2]

10-element Vector{Int64}:
 19
 16
 19
 16
 18
 18
 13
 11
 20
 19

In [69]:
# выбор всех значений в столбцах 2 и 5:
ar[:, [2, 5]]

10×2 Matrix{Int64}:
 19  15
 16  12
 19  14
 16  12
 18  18
 18  12
 13  13
 11  18
 20  11
 19  19

In [70]:
# все значения строк в столбцах 2, 3 и 4:
ar[:, 2:4]

10×3 Matrix{Int64}:
 19  11  16
 16  11  17
 19  20  12
 16  20  15
 18  10  15
 18  16  17
 13  18  10
 11  12  12
 20  17  20
 19  10  20

In [71]:
# значения в строках 2, 4, 6 и в столбцах 1 и 5:
ar[[2, 4, 6], [1, 5]]

3×2 Matrix{Int64}:
 12  12
 11  12
 17  12

In [72]:
# значения в строке 1 от столбца 3 до последнего столбца:
ar[1, 3:end]

3-element Vector{Int64}:
 11
 16
 15

In [73]:
# сортировка по столбцам:
sort(ar,dims=1)

10×5 Matrix{Int64}:
 11  11  10  10  11
 12  13  10  12  12
 12  16  11  12  12
 12  16  11  15  12
 14  18  12  15  13
 16  18  16  16  14
 17  19  17  17  15
 18  19  18  17  18
 18  19  20  20  18
 19  20  20  20  19

In [74]:
# сортировка по строкам:
sort(ar,dims=2)

10×5 Matrix{Int64}:
 11  15  16  18  19
 11  12  12  16  17
 12  12  14  19  20
 11  12  15  16  20
 10  14  15  18  18
 12  16  17  17  18
 10  13  13  18  19
 11  12  12  18  18
 11  12  17  20  20
 10  16  19  19  20

In [75]:
# поэлементное сравнение с числом
# (результат - массив логических значений):
ar .> 14

10×5 BitMatrix:
 1  1  0  1  1
 0  1  0  1  0
 0  1  1  0  0
 0  1  1  1  0
 0  1  0  1  1
 1  1  1  1  0
 1  0  1  0  0
 1  0  0  0  1
 0  1  1  1  0
 1  1  0  1  1

In [76]:
# возврат индексов элементов массива, удовлетворяющих условию:
findall(ar .> 14)

29-element Vector{CartesianIndex{2}}:
 CartesianIndex(1, 1)
 CartesianIndex(6, 1)
 CartesianIndex(7, 1)
 CartesianIndex(8, 1)
 CartesianIndex(10, 1)
 CartesianIndex(1, 2)
 CartesianIndex(2, 2)
 CartesianIndex(3, 2)
 CartesianIndex(4, 2)
 CartesianIndex(5, 2)
 CartesianIndex(6, 2)
 CartesianIndex(9, 2)
 CartesianIndex(10, 2)
 ⋮
 CartesianIndex(9, 3)
 CartesianIndex(1, 4)
 CartesianIndex(2, 4)
 CartesianIndex(4, 4)
 CartesianIndex(5, 4)
 CartesianIndex(6, 4)
 CartesianIndex(9, 4)
 CartesianIndex(10, 4)
 CartesianIndex(1, 5)
 CartesianIndex(5, 5)
 CartesianIndex(8, 5)
 CartesianIndex(10, 5)

## Самостоятельное задание

### 1. Даны множества: 𝐴 = {0, 3, 4, 9}, 𝐵 = {1, 3, 4, 7}, 𝐶 = {0, 1, 2, 4, 7, 8, 9}. Найти 𝑃 = 𝐴 ∩ 𝐵 ∪ 𝐴 ∩ 𝐵 ∪ 𝐴 ∩ 𝐶 ∪ 𝐵 ∩ C.

In [77]:
A = Set([0, 3, 4, 9])
B = Set([1, 3, 4, 7])
C = Set([0, 1, 2, 4, 7, 8, 9])

Set{Int64} with 7 elements:
  0
  4
  7
  2
  9
  8
  1

In [91]:
P = A ∩ B ∪ A ∩ B ∪ A ∩ C ∪ B ∩ C

Set{Int64} with 6 elements:
  0
  4
  7
  9
  3
  1

In [92]:
P1 = A ∩ (B ∪ (A ∩ (B ∪ (A ∩ (C ∪ (B ∩ C))))))

Set{Int64} with 4 elements:
  0
  4
  9
  3

In [94]:
print(P), print(P1)

Set([0, 4, 7, 9, 3, 1])Set([0, 4, 9, 3])

(nothing, nothing)

### 2. Приведите свои примеры с выполнением операций над множествами элементов разных типов.

In [95]:
A1 = Set("ajkhewrhglewha")

Set{Char} with 9 elements:
  'w'
  'a'
  'e'
  'j'
  'h'
  'g'
  'k'
  'r'
  'l'

In [96]:
B1 = Set([1 2 3 4 5 6])

Set{Int64} with 6 elements:
  5
  4
  6
  2
  3
  1

In [97]:
C1 = Set(["wgegetr", "agsegqe", "wggertgert"])

Set{String} with 3 elements:
  "wgegetr"
  "wggertgert"
  "agsegqe"

In [98]:
A1 ∩ B1

Set{Any}()

In [99]:
A1 ∪ B1

Set{Any} with 15 elements:
  5
  'w'
  'a'
  'e'
  1
  4
  6
  2
  'j'
  'h'
  'g'
  'k'
  'r'
  'l'
  3

In [101]:
setdiff(A1, C1)

Set{Char} with 9 elements:
  'w'
  'a'
  'e'
  'j'
  'h'
  'g'
  'k'
  'r'
  'l'

In [102]:
issubset(A1, B1)

false

### 3. Создайте разными способами:

In [104]:
N = 50;

3.1) массив (1, 2, 3, … 𝑁 − 1, 𝑁 ), 𝑁 выберите больше 20;

In [126]:
B31 = [i for i ∈ range(1, N)];
C31 = collect(1:N);
# Классическим циклом
#=D31 = (Int64)[]
for i ∈ 1:N
    push!(D31, i)
end
print(D31)=#
A31 = [i for i ∈ 1:N]

50-element Vector{Int64}:
  1
  2
  3
  4
  5
  6
  7
  8
  9
 10
 11
 12
 13
  ⋮
 39
 40
 41
 42
 43
 44
 45
 46
 47
 48
 49
 50

3.2) массив (𝑁, 𝑁 − 1 … , 2, 1), 𝑁 выберите больше 20;

In [127]:
B32 = [i for i ∈ range(N, 1, step=-1)];
C32 = collect(N:-1:1);
# Классическим циклом
#=D32 = (Int64)[]
for i ∈ 1:N
    push!(D32, i)
end
print(D32)=#
A32 = [i for i ∈ N:-1:1]

50-element Vector{Int64}:
 50
 49
 48
 47
 46
 45
 44
 43
 42
 41
 40
 39
 38
  ⋮
 12
 11
 10
  9
  8
  7
  6
  5
  4
  3
  2
  1

3.3) массив (1, 2, 3, … , 𝑁 − 1, 𝑁, 𝑁 − 1, … , 2, 1), 𝑁 выберите больше 20;

In [131]:
B33_1 = [i for i ∈ range(1, N)];
B33_2 = [i for i ∈ range(N-1, 1, step=-1)];
B33 = [B33_1; B33_2];
C33 = vcat(B33_1, B33_2);
#C32 = collect(N:-1:1);
# Классическим циклом
D33 = (Int64)[]
for i ∈ 1:N
    push!(D33, i)
end
for i ∈ N-1:-1:1
    push!(D33, i)
end
print(D33)
#A32 = [i for i ∈ N:-1:1]

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 49, 48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

3.4) массив с именем tmp вида (4, 6, 3);

In [132]:
tmp = [4, 6, 3]

3-element Vector{Int64}:
 4
 6
 3

3.5) массив, в котором первый элемент массива tmp повторяется 10 раз;

In [139]:
B35 = fill(tmp[1], 10);
C35 = repeat([tmp[1]], 10);
A35 = [tmp[1] for i in 1:10]

10-element Vector{Int64}:
 4
 4
 4
 4
 4
 4
 4
 4
 4
 4

3.6) массив, в котором все элементы массива tmp повторяются 10 раз;

In [151]:
B36 = fill(tmp, 10);
C36_1 = repeat(tmp, 10); # В один столбец (30 x 1)

In [152]:
C36_2 = repeat(tmp, 1, 10) # В строки (3 x 10)

3×10 Matrix{Int64}:
 4  4  4  4  4  4  4  4  4  4
 6  6  6  6  6  6  6  6  6  6
 3  3  3  3  3  3  3  3  3  3

In [153]:
A36 = [tmp for i in 1:10]

10-element Vector{Vector{Int64}}:
 [4, 6, 3]
 [4, 6, 3]
 [4, 6, 3]
 [4, 6, 3]
 [4, 6, 3]
 [4, 6, 3]
 [4, 6, 3]
 [4, 6, 3]
 [4, 6, 3]
 [4, 6, 3]

3.7) массив, в котором первый элемент массива tmp встречается 11 раз, второй элемент — 10 раз, третий элемент — 10 раз;

In [155]:
B37_1 = fill(tmp[1], 11);
B37_2 = fill(tmp[2], 10);
B37_3 = fill(tmp[3], 10);
B37 = [B37_1; B37_2; B37_3]

31-element Vector{Int64}:
 4
 4
 4
 4
 4
 4
 4
 4
 4
 4
 4
 6
 6
 ⋮
 6
 6
 3
 3
 3
 3
 3
 3
 3
 3
 3
 3

In [157]:
A37 = [tmp[i % 3 + 1] for i ∈ 0:30]

31-element Vector{Int64}:
 4
 6
 3
 4
 6
 3
 4
 6
 3
 4
 6
 3
 4
 ⋮
 6
 3
 4
 6
 3
 4
 6
 3
 4
 6
 3
 4

3.8) массив, в котором первый элемент массива tmp встречается 10 раз подряд, второй
элемент — 20 раз подряд, третий элемент — 30 раз подряд;

In [158]:
B38_1 = fill(tmp[1], 10);
B38_2 = fill(tmp[2], 20);
B38_3 = fill(tmp[3], 30);
B38 = [B38_1; B38_2; B38_3]

60-element Vector{Int64}:
 4
 4
 4
 4
 4
 4
 4
 4
 4
 4
 6
 6
 6
 ⋮
 3
 3
 3
 3
 3
 3
 3
 3
 3
 3
 3
 3

In [169]:
A38 = (Int64)[]
for i in 1:10
    push!(A38, tmp[1])
end
for i in 1:20
    push!(A38, tmp[2])
end
for i in 1:30
    push!(A38, tmp[3])
end
A38

60-element Vector{Int64}:
 4
 4
 4
 4
 4
 4
 4
 4
 4
 4
 6
 6
 6
 ⋮
 3
 3
 3
 3
 3
 3
 3
 3
 3
 3
 3
 3

3.9) массив из элементов вида $2^{𝑡𝑚𝑝[𝑖]}$, $𝑖 = 1, 2, 3$, где элемент $2^{𝑡𝑚𝑝[3]}$ встречается 4
раза; посчитайте в полученном векторе, сколько раз встречается цифра 6, и выведите это значение на экран;

In [189]:
A39 = [2^(tmp[i % 3 + 1]) for i ∈ 0:11]

12-element Vector{Int64}:
 16
 64
  8
 16
 64
  8
 16
 64
  8
 16
 64
  8

In [191]:
function digit_num(d, a)
   # Находит число цифр d в числе a
    n = 0
    while a > 0
        if a % 10 == d
            n += 1
        end
        a = a ÷ 10
    end
    return n
end

digit_num (generic function with 1 method)

In [198]:
N1 = sum([digit_num(6, a) for a in A39]);
N2 = reduce(+, [digit_num(6, a) for a in A39])

8

3.10) вектор значений $y = e^x \cos(x)$ в точках $x = 3, 3.1, 3.2, \dots , 6$, найдите среднее
значение $y$;

In [203]:
A310 = [exp(x)*cos(x) for x ∈ 3:0.1:6]

31-element Vector{Float64}:
 -19.884530844146987
 -22.178753389342127
 -24.490696732801293
 -26.77318244299338
 -28.969237768093574
 -31.011186439374516
 -32.819774760338504
 -34.30336011037369
 -35.35719361853035
 -35.86283371230767
 -35.68773248011913
 -34.68504225166807
 -32.693695428321746
   ⋮
  25.046704998273004
  42.09920106253839
  61.99663027669454
  84.92906736250268
 111.0615860420258
 140.5250750527875
 173.40577640857734
 209.73349424783467
 249.46844055885668
 292.4867067371223
 338.5643778585117
 387.36034029093076

In [204]:
Avg_310 = sum(A310)/length(A310)

53.11374594642971

3.11) вектор вида $(x^i, y^j)$, $x = 0.1$, $i = 3, 6, 9, \dots, 36$, $y = 0.2$, $j = 1, 4, 7, \dots, 34$;

In [209]:
x = 0.1
y = 0.2
A311 = [(x^i, y^j) for i ∈ 3:3:36 for j ∈ 1:3:34]  # j - внутренний цикл

144-element Vector{Tuple{Float64, Float64}}:
 (0.0010000000000000002, 0.2)
 (0.0010000000000000002, 0.0016000000000000003)
 (0.0010000000000000002, 1.2800000000000006e-5)
 (0.0010000000000000002, 1.0240000000000006e-7)
 (0.0010000000000000002, 8.192000000000005e-10)
 (0.0010000000000000002, 6.5536000000000055e-12)
 (0.0010000000000000002, 5.2428800000000056e-14)
 (0.0010000000000000002, 4.194304000000005e-16)
 (0.0010000000000000002, 3.3554432000000044e-18)
 (0.0010000000000000002, 2.684354560000004e-20)
 (0.0010000000000000002, 2.147483648000004e-22)
 (0.0010000000000000002, 1.7179869184000035e-24)
 (1.0000000000000004e-6, 0.2)
 ⋮
 (1.000000000000002e-36, 0.2)
 (1.000000000000002e-36, 0.0016000000000000003)
 (1.000000000000002e-36, 1.2800000000000006e-5)
 (1.000000000000002e-36, 1.0240000000000006e-7)
 (1.000000000000002e-36, 8.192000000000005e-10)
 (1.000000000000002e-36, 6.5536000000000055e-12)
 (1.000000000000002e-36, 5.2428800000000056e-14)
 (1.000000000000002e-36, 4.1943040000000

3.12) вектор с элементами $\frac{2^i}{i}$, $i = 1, 2, \dots, M$, $M = 25$;

In [210]:
M = 25
A312 = [2^i / i for i in 1:M]

25-element Vector{Float64}:
      2.0
      2.0
      2.6666666666666665
      4.0
      6.4
     10.666666666666666
     18.285714285714285
     32.0
     56.888888888888886
    102.4
    186.1818181818182
    341.3333333333333
    630.1538461538462
   1170.2857142857142
   2184.5333333333333
   4096.0
   7710.117647058823
  14563.555555555555
  27594.105263157893
  52428.8
  99864.38095238095
 190650.18181818182
 364722.0869565217
 699050.6666666666
      1.34217728e6

3.13) вектор вида (”fn1”, ”fn2”, …, ”fnN”), $𝑁 = 30$;

In [212]:
N = 30
A313 = ["fn$i" for i in 1:N]

30-element Vector{String}:
 "fn1"
 "fn2"
 "fn3"
 "fn4"
 "fn5"
 "fn6"
 "fn7"
 "fn8"
 "fn9"
 "fn10"
 "fn11"
 "fn12"
 "fn13"
 ⋮
 "fn19"
 "fn20"
 "fn21"
 "fn22"
 "fn23"
 "fn24"
 "fn25"
 "fn26"
 "fn27"
 "fn28"
 "fn29"
 "fn30"

3.14) векторы $𝑥 = (𝑥_1, 𝑥_2, … , 𝑥_𝑛)$ и $𝑦 = (𝑦_1, 𝑦_2, … , 𝑦_𝑛)$ целочисленного типа длины $𝑛 = 250$ как случайные выборки из совокупности $0, 1, … , 999$; 
на его основе:

In [223]:
x = [rand(0:999) for i in 1:n]

250-element Vector{Int64}:
 304
 903
 949
 531
 245
 968
 953
 414
 692
 717
  46
 582
 146
   ⋮
 919
 718
 407
 231
  59
 246
 756
 199
 322
 939
 498
 297

In [224]:
y = [rand(0:999) for i in 1:n]

250-element Vector{Int64}:
 752
 859
  49
 809
 904
 814
 511
 168
  40
 411
 447
  86
 448
   ⋮
 987
 285
 804
 612
 602
 410
  36
 677
 887
 614
  92
 391

– сформируйте вектор $(𝑦_2 − 𝑥_1, \dots, 𝑦_𝑛 − 𝑥_{𝑛−1})$;

In [225]:
y_minus_x = [y[i] - x[i-1] for i ∈ 2:n]

249-element Vector{Int64}:
  555
 -854
 -140
  373
  569
 -457
 -785
 -374
 -281
 -270
   40
 -134
  761
    ⋮
  237
 -634
   86
  205
  371
  351
 -210
  -79
  688
  292
 -847
 -107

– сформируйте вектор $(𝑥_1 + 2𝑥_2 − 𝑥_3, 𝑥_2 + 2𝑥_3 − 𝑥_4, \dots, 𝑥_{𝑛−2} + 2𝑥_{𝑛−1} − 𝑥_𝑛)$;

In [226]:
x1_plus_2x2_minus_x3 = [x[i] + 2*x[i+1] - x[i+2] for i ∈ 1:n-2]

248-element Vector{Int64}:
 1161
 2270
 1766
   53
 1228
 2460
 1089
 1081
 2080
  227
 1064
   28
 1026
    ⋮
 1245
 1870
 1948
 1301
  810
  103
 -205
 1559
  832
  -96
 1702
 1638

– сформируйте вектор $(\frac{\sin(y_1)}{\cos(x_2)}, \frac{\sin(y_2)}{\cos(x_3)}, \dots, \frac{\sin(y_{n-1})}{\cos(x_n)})$

In [227]:
siny_divide_cosx = [sin(y[i-1]) / cos(x[i]) for i ∈ 2:n]

249-element Vector{Float64}:
   4.440762796604472
  -1.0031818070105352
   0.9561508827440894
  -1.000182463515962
  -0.7591972657062622
   0.7056098258205616
   1.1435236257292338
  -1.5102475163473301
   0.9883937960184693
  -1.2067373827245433
  -1.1250376112108684
 -10.998864247308298
  -1.5481986411729924
   ⋮
  11.86408519204383
  -3.5315825458429986
   4.747080802129639
  -2.641994690524218
  -0.743544523690401
  -1.606294475559757
  -2.3130027041569248
   2.1028939413502497
 -75.48444446193533
  -0.9296670498176479
  17.096364201961297
   6.538592779718017

– вычислите $\sum\limits_{i = 1}^{n-1} \frac{e^{-x_{i+1}}}{x_i + 10}$

In [228]:
S314_array = [exp(-x[i+1])/(x[i]+10) for i ∈ 1:n-1]

249-element Vector{Float64}:
 0.0
 0.0
 2.5574756191608085e-234
 7.322420557008654e-110
 0.0
 0.0
 1.6537056299290693e-183
 6.931906024237895e-304
 5.81459356e-315
 1.448503075316893e-23
 3.1075884977133967e-255
 6.617347406166209e-67
 0.0
 ⋮
 0.0
 1.616390463e-315
 2.3989129343706285e-180
 1.1424537443123088e-103
 9.876624102466392e-29
 2.1120702587699017e-109
 0.0
 4.910993186809479e-90
 6.871167586659815e-143
 0.0
 5.547301870445639e-220
 2.0355190089610133e-132

In [229]:
S314 = sum(S314_array)

0.0029792248348674754

– выберите элементы вектора 𝑦, значения которых больше 600, и выведите на экран; определите индексы этих элементов;

In [235]:
y_bigger_than_600 = [y_ for y_ ∈ y if y_ > 600]

113-element Vector{Int64}:
 752
 859
 809
 904
 814
 907
 664
 629
 778
 870
 690
 835
 622
   ⋮
 759
 837
 852
 630
 887
 987
 804
 612
 602
 677
 887
 614

In [236]:
indexes_y_bigger_than_600 = findall(y .> 600)

113-element Vector{Int64}:
   1
   2
   4
   5
   6
  14
  15
  16
  17
  18
  19
  20
  21
   ⋮
 227
 229
 232
 233
 235
 239
 241
 242
 243
 246
 247
 248

– определите значения вектора 𝑥, соответствующие значениям вектора 𝑦, значения которых больше 600 (под соответствием понимается расположение на
аналогичных индексных позициях);

In [238]:
x_with_indexes_as_y_bigger_than_600 = [x[i] for i ∈ indexes_y_bigger_than_600]

113-element Vector{Int64}:
 304
 903
 531
 245
 968
 846
 812
 204
  57
   7
 339
 618
 913
   ⋮
 976
 239
 133
 787
 274
 919
 407
 231
  59
 199
 322
 939

– сформируйте вектор $(|𝑥_1 − \overline{x}|, |𝑥_2 − \overline{x}|, \dots, |𝑥_n − \overline{x}|)$, где $\overline{𝑥}$ обозначает
среднее значение вектора $x = (x_1, x_2, \dots, x_n)$

In [239]:
avg_x = sum(x)/length(x)

500.776

In [240]:
abs_x_minus_avgx = [abs(x[i] - avg_x) for i ∈ 1:n]

250-element Vector{Float64}:
 196.776
 402.224
 448.224
  30.22399999999999
 255.776
 467.224
 452.224
  86.77600000000001
 191.224
 216.224
 454.776
  81.22399999999999
 354.776
   ⋮
 418.224
 217.224
  93.77600000000001
 269.776
 441.776
 254.776
 255.224
 301.776
 178.776
 438.224
   2.7760000000000105
 203.776

– определите, сколько элементов вектора 𝑦 отстоят от максимального значения не более, чем на 200;

In [241]:
y_max = maximum(y)

993

In [242]:
y_dist_less_than_maxy_by200 = [y_ for y_ in y if abs(y_ - y_max) <= 200]

54-element Vector{Int64}:
 859
 809
 904
 814
 907
 870
 835
 852
 803
 924
 798
 881
 866
   ⋮
 952
 836
 934
 978
 941
 950
 837
 852
 887
 987
 804
 887

In [243]:
number_of_y_dist_less_than_maxy_by200 = length(y_dist_less_than_maxy_by200)

54

– определите, сколько чётных и нечётных элементов вектора 𝑥;

In [246]:
number_of_odd_x = count(i->(i % 2 == 1), x)

130

In [248]:
number_of_even_x = length([i for i ∈ x if i % 2 == 0])

120

– определите, сколько элементов вектора 𝑥 кратны 7;

In [249]:
number_of_x_divisible_by_7 = length([x_ for x_ ∈ x if x_ % 7 == 0])

39

– отсортируйте элементы вектора 𝑥 в порядке возрастания элементов вектора 𝑦;

In [255]:
y_ascending_order_indexes = sortperm(y) # Индексы отсортированного по возрастанию массива y

250-element Vector{Int64}:
  53
 174
 170
  29
 237
 245
   9
 214
  90
 130
 223
  22
   3
   ⋮
 224
  57
 225
 185
 189
 138
 211
  78
 145
 239
 102
 120

In [262]:
#x_sorted_by_indexes_of_y_ascending_order = [x[i] for i ∈ y_ascending_order_indexes]
x_sorted_by_indexes_of_y_ascending_order = x[sortperm(y)]

250-element Vector{Int64}:
 307
  40
 330
 285
 664
 756
 692
 102
  96
 710
 831
 733
 949
   ⋮
 251
 704
 350
 332
 739
 593
 993
 700
 294
 919
 769
 884

– выведите элементы вектора 𝑥, которые входят в десятку наибольших (top-10)

In [258]:
top_10_x = sort(x)[end-9:end]

10-element Vector{Int64}:
 968
 970
 976
 977
 984
 985
 989
 993
 994
 995

– сформируйте вектор, содержащий только уникальные (неповторяющиеся) элементы вектора $x$

In [259]:
x_unique = unique(x)

228-element Vector{Int64}:
 304
 903
 949
 531
 245
 968
 953
 414
 692
 717
  46
 582
 146
   ⋮
 919
 718
 407
 231
  59
 246
 756
 199
 322
 939
 498
 297

### 4. Создайте массив squares, в котором будут храниться квадраты всех целых чисел от 1 до 100

In [260]:
squares = [i^2 for i ∈ 1:100]

100-element Vector{Int64}:
     1
     4
     9
    16
    25
    36
    49
    64
    81
   100
   121
   144
   169
     ⋮
  7921
  8100
  8281
  8464
  8649
  8836
  9025
  9216
  9409
  9604
  9801
 10000

### 5. Подключите пакет Primes (функции для вычисления простых чисел)

Сгенерируйте
массив myprimes, в котором будут храниться первые 168 простых чисел. Определите
89-е наименьшее простое число. Получите срез массива с 89-го до 99-го элемента
включительно, содержащий наименьшие простые числа.

In [261]:
import Pkg
Pkg.add("Primes")
using Primes

    Updating registry at `C:\Users\User\.julia\registries\General.toml`
   Resolving package versions...
   Installed Primes ─────────── v0.5.5
   Installed IntegerMathUtils ─ v0.1.2
    Updating `C:\Users\User\.julia\environments\v1.8\Project.toml`
  [27ebfcd6] + Primes v0.5.5
    Updating `C:\Users\User\.julia\environments\v1.8\Manifest.toml`
  [18e54dd8] + IntegerMathUtils v0.1.2
  [27ebfcd6] + Primes v0.5.5
Precompiling project...
  ✓ IntegerMathUtils
  ✓ Primes
  2 dependencies successfully precompiled in 31 seconds. 285 already precompiled.


In [264]:
myprimes = primes(1000)

168-element Vector{Int64}:
   2
   3
   5
   7
  11
  13
  17
  19
  23
  29
  31
  37
  41
   ⋮
 919
 929
 937
 941
 947
 953
 967
 971
 977
 983
 991
 997

In [265]:
myprimes[89]

461

In [266]:
myprimes[89:99]

11-element Vector{Int64}:
 461
 463
 467
 479
 487
 491
 499
 503
 509
 521
 523

### 6. Вычислите следующие выражения:

6.1) $\sum\limits_{i=10}^{100} (i^3 + 4i^2)$

In [267]:
S61 = sum([i^3 + 4*i^2 for i ∈ 10:100])

26852735

6.2) $\sum\limits_{i=1}^{M} (\frac{2^i}{i} + \frac{3^i}{i^2})$, где $M = 25$

In [270]:
S62 = reduce(+, [2^i / i + 3^i / i^2 for i ∈ 1:M])

2.1291704368143802e9

6.3) $1 + \frac{2}{3} + (\frac{2}{3}\frac{4}{5}) + (\frac{2}{3}\frac{4}{5}\frac{6}{7}) + \cdots + (\frac{2}{3}\frac{4}{5} \cdots \frac{38}{39})$

In [279]:
[1 - 1//i for i ∈ 3:2:39]

19-element Vector{Rational{Int64}}:
  2//3
  4//5
  6//7
  8//9
 10//11
 12//13
 14//15
 16//17
 18//19
 20//21
 22//23
 24//25
 26//27
 28//29
 30//31
 32//33
 34//35
 36//37
 38//39

In [280]:
cumprod([1 - 1//i for i ∈ 3:2:39])

19-element Vector{Rational{Int64}}:
           2//3
           8//15
          16//35
         128//315
         256//693
        1024//3003
        2048//6435
       32768//109395
       65536//230945
      262144//969969
      524288//2028117
     4194304//16900975
     8388608//35102025
    33554432//145422675
    67108864//300540195
  2147483648//9917826435
  4294967296//20419054425
 17179869184//83945001525
 34359738368//172308161025

In [288]:
S63 = 1 + reduce(+, cumprod([1 - 1/i for i ∈ 3:2:39]))

6.976346137897621